omnipaxos_core/src/sequence_paxos/mod.rs

```diff
@@ -217,37 +218,60 @@ where
-                self.send_prepare(peer);
-            }
-        }
-
-        // Resend AcceptStopSign
-        if self.state == (Role::Leader, Phase::Accept) {
-            // TODO: This is slow. Get stopsign from cache instead.
-            match self.internal_storage.get_stopsign() {
-                Some(ss) if !ss.decided => {
-                    for follower in self.leader_state.get_promised_followers() {
-                        if !self.leader_state.get_accepted_stopsign(follower) {
-                            // TODO: subtract 1 from seq_num to show duplicated?
-                            self.send_accept_stopsign(follower, ss.stopsign.clone());
+        match &self.state {
+            (Role::Leader, phase) => {
+                // Resend AcceptStopSign
+                if *phase == Phase::Accept {
+                    // TODO: This is slow. Get stopsign from cache instead.
+                    if let Some(ss) = self.internal_storage.get_stopsign() {
+                        if !ss.decided {
+                            for follower in self.leader_state.get_promised_followers() {
+                                if !self.leader_state.follower_has_accepted_stopsign(follower) {
+                                    self.send_accept_stopsign(follower, ss.stopsign.clone(), true);
+                                }
+                            }
                         }
                     }
                 }
-                _ => (),
-            }
-        }
 
-        // Resend PrepareReq
-        if self.state == (Role::Follower, Phase::Recover) {
-            self.outgoing.push(PaxosMessage {
-                from: self.pid,
-                to: self.leader.pid,
-                msg: PaxosMsg::PrepareReq,
-            });
+                // Resend Prepare
+                let unpromised_peers = self.leader_state.get_unpromised_peers();
+                for peer in unpromised_peers {
+                    self.send_prepare(peer);
+                }
+            },
+            (Role::Follower, phase) => {
+                match phase {
+                    Phase::Recover => {
+                        // Resend PrepareReq
+                        self.outgoing.push(PaxosMessage {
+                            from: self.pid,
+                            to: self.leader.pid,
+                            msg: PaxosMsg::PrepareReq,
+                        });
+                    }
+                    Phase::Prepare => {
+                        // Resend Promise
+                        match &self.cached_promise {
+                            Some(promise) => {
+                                self.outgoing.push(PaxosMessage {
+                                    from: self.pid,
+                                    to: promise.n.pid,
+                                    msg: PaxosMsg::Promise(promise.clone()),
+                                });
+                            },
+                            None => {
```

Comment:
This shouldn't happen, but if it really does, perhaps it makes sense to try and recover this server by going into recovery mode and sending PrepareReq to all peers. This will also help to avoid excessive logging.